website/static/js/fileBrowser.js

```diff
@@ -0,0 +1,1175 @@
+        }
+    }
+    return root.children;
+}
+
+/**
+ * Returns the object to send to the API to end a node_link to collection
+ * @param id {String} unique id of the node like 'ez8f3'
+ * @returns {{data: {type: string, relationships: {nodes: {data: {type: string, id: *}}}}}}
+ */
+function buildCollectionNodeData (id) {
+    return {
+        'data' : {
+            'type': 'node_links',
+            'relationships': {
+                'nodes': {
+                    'data': {
+                        'type': 'nodes',
+                        'id': id
+                    }
+                }
+            }
+        }
+    };
+}
+
+/**
+ * Initialize File Browser. Prepeares an option object within FileBrowser
+ * @constructor
+ */
+var FileBrowser = {
+    controller : function (options) {
+        var self = this;
+        self.wrapperSelector = options.wrapperSelector;  // For encapsulating each implementation of this component in multiple use
+        self.currentLink = ''; // Save the link to compare if a new link is being requested and avoid multiple calls
+        self.reload = m.prop(false); // Gets set to true when treebeard link changes and it needs to be redrawn
+        self.nonLoadTemplate = m.prop(m('.fb-non-load-template.m-md.p-md.osf-box', 'Loading...')); // Template for when data is not available or error happens
+
+        // VIEW STATES
+        self.showInfo = m.prop(true); // Show the info panel
+        self.showSidebar = m.prop(true); // Show the links with collections etc. used in narrow views
+        self.refreshView = m.prop(true); // Internal loading indicator
+        self.allProjectsLoaded = m.prop(false);
+        self.allProjects = m.prop([]);
+        self.loadingNodePages = false; // Since API returns pages of items, this state shows whether filebrowser is still loading the next pages.
+        self.loadingAllNodes = false; // True if we are loading all nodes
+
+        // Load 'All my Projects' and 'All my Registrations'
+        self.systemCollections = [
+            new LinkObject('collection', { path : 'users/me/nodes/', query : { 'related_counts' : true, 'page[size]'  : 12, 'embed' : 'contributors' }, systemCollection : 'nodes'}, 'All My Projects'),
+            new LinkObject('collection', { path : 'users/me/registrations/', query : { 'related_counts' : true, 'page[size]'  : 12, 'embed' : 'contributors'}, systemCollection : 'registrations'}, 'All My Registrations')
+        ];
+        // Initial Breadcrumb for All my projects
+        self.breadcrumbs = m.prop([
+            new LinkObject('collection', { path : 'users/me/nodes/', query : { 'related_counts' : true, 'embed' : 'contributors' }, systemCollection : 'nodes'}, 'All My Projects')
+        ]);
+        // Calculate name filters
+        self.nameFilters = [];
+        // Calculate tag filters
+        self.tagFilters = [];
+
+        // Placeholder for node data
+        self.data = m.prop([]);
```

Comment:
`self.data` is initialized as an Array prop, but it is treated elsewhere in the code as an object, which is confusing. For example, `self.data().data` is assigned in `updateListError`. Better to ensure that `self.data` is always an Array, unwrapping the `data` envelope when necessary.